hibernate-cache-local/src/main/java/org/infinispan/tutorial/simple/hibernate/cache/local/InfinispanHibernateCacheLocal.java

```diff
@@ -0,0 +1,268 @@
+   }
+
+   private static void updateEntity(long id) {
+      try(Session em = createEntityManagerWithStatsCleared()) {
+         EntityTransaction tx = em.getTransaction();
+         try {
+            tx.begin();
+
+            Event event = em.find(Event.class, id);
+            event.setName("Caught a Snorlax!!");
+
+            System.out.printf("Updated entity: %s%n", event);
+         } catch (Throwable t) {
+            tx.setRollbackOnly();
+            throw t;
+         } finally {
+            if (tx.isActive()) tx.commit();
+            else tx.rollback();
+         }
+      }
+   }
+
+   private static void evictEntity(long id) {
+      try(Session em = createEntityManagerWithStatsCleared()) {
+         em.getEntityManagerFactory().getCache().evict(Event.class, id);
+      }
+   }
+
+   private static void deleteEntity(long id) {
+      try(Session em = createEntityManagerWithStatsCleared()) {
+         EntityTransaction tx = em.getTransaction();
+         try {
+            tx.begin();
+
+            Event event = em.find(Event.class, id);
+            em.remove(event);
+         } catch (Throwable t) {
+            tx.setRollbackOnly();
+            throw t;
+         } finally {
+            if (tx.isActive()) tx.commit();
+            else tx.rollback();
+         }
+      }
+   }
+
+   private static void queryEntities() {
+      try(Session em = createEntityManagerWithStatsCleared()) {
+         TypedQuery<Event> query = em.createQuery("from Event", Event.class);
+         query.setHint("org.hibernate.cacheable", Boolean.TRUE);
+         List<Event> events = query.getResultList();
+         System.out.printf("Queried events: %s%n", events);
+      }
+   }
+
+   private static void saveExpiringEntity() {
+      try(Session em = createEntityManagerWithStatsCleared()) {
+         EntityTransaction tx = em.getTransaction();
+         try {
+            tx.begin();
+
+            em.persist(new Person("Satoshi"));
+         } catch (Throwable t) {
+            tx.setRollbackOnly();
+            throw t;
+         } finally {
+            if (tx.isActive()) tx.commit();
+            else tx.rollback();
+         }
+      }
+   }
+
+   private static void findExpiringEntity(long id) {
+      try(Session em = createEntityManagerWithStatsCleared()) {
+         Person person = em.find(Person.class, id);
+         System.out.printf("Found expiring entity: %s%n", person);
+      }
+   }
+
+   private static SecondLevelCacheStatistics getCacheStatistics(String regionName) {
+      return emf.unwrap(SessionFactory.class).getStatistics()
+            .getSecondLevelCacheStatistics(regionName);
+   }
+
+   private static Statistics getStatistics() {
+      return emf.unwrap(SessionFactory.class).getStatistics();
+   }
+
+   private static void printfAssert(String format, long actual, int expected) {
+      System.out.printf(format, actual, expected);
+      assert expected == actual : errorMsgAndCloseEntityManagerFactory(actual, expected);
```

Comment:
Side effect (closing emf) from assert message looks clunky... use rather if ... throw some exception

Reply:
Makes sense. Moving the code to simply throw exceptions and hence does not need assertions any more.